lib/db.js

```diff
@@ -84,7 +103,93 @@ class LimitDBRedis extends EventEmitter {
     });
   }
 
+  #setupPing() {
+    this.failedPings = 0;
+    this.redis.on('ready', () => this.#startPing())  
+  }
+
+  #startPing() {
+    if (this.keepPingRunning) {
+      this.#stopPing()
+    }
+
+    this.keepPingRunning = true;
+    const doPing = () => {
+      if (!this.pingConfig?.enabled()) {
+        setTimeout(doPing, this.pingConfig.msBetweenEnabledChecks);
+        return;
+      }
+      let start = Date.now();
+      this.redis.ping((err) => {
+        let duration = Date.now()-start;
+        err ? this.#pingKO(triggerLoop, err, duration) : this.#pingOK(triggerLoop, duration)
+      });
+    }
+
+    const triggerLoop = () => {
+      if(this.keepPingRunning) {
+        this.pingIntervalId = setTimeout(doPing, this.pingConfig.interval);
+      }
+    }
+
+    doPing();
+  }
+
+  #pingOK(callback, duration) {
+    this.reconnecting = false
+    this.failedPings = 0;
+    this.#emitPingResult(PING_SUCCESS, undefined, duration, 0)
+    callback();
+  }
+
+  #pingKO(callback, err, duration) {
+    this.failedPings++;
+    this.#emitPingResult(PING_ERROR, err, duration, this.failedPings)
+
+    if (this.failedPings >= this.pingConfig.maxFailedAttempts && 
+        (!this.reconnecting || this.failedPings % this.pingConfig.maxFailedAttemptsToRetryReconnect == 0)) {
+          if(!this.pingConfig.reconnectIfFailed() ) {
+            this.#emitPingResult(PING_RECONNECT_DRY_RUN, undefined, 0, this.failedPings)
+            callback()
+          } else {
+            this.reconnecting = true
```

Comment:
what's the semantic behind `reconnecting` and `keepPingRunning` they look pretty similar to me